styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -0,0 +1,241 @@
+ * stored state.
+ */
+public class ShardedCounter {
+
+  // Ought to be enough (parallelism) for everyone. We could make it dynamic with extra effort.
+  private static final int NUM_SHARDS = 128;
+  private static final long CACHE_EXPIRY_MILLIS = 1000;
+
+  private final Datastore datastore;
+  public static final String KIND_COUNTER_LIMIT = "CounterLimit";
+  public static final String PROPERTY_LIMIT = "limit";
+  public static final String KIND_COUNTER_SHARD = "CounterShard";
+  public static final String PROPERTY_SHARD = "shard";
+
+  private static class CounterSnapshot {
+    Instant updatedAt;
+    Long limit;
+    ArrayList<Long> shards;
+
+    /**
+     * Idempotent initialization, so that we don't reset an existing shard to zero - counterId may
+     * have already been initialized and incremented by another process.
+     */
+    private static void initialize(Datastore datastore, String counterId) {
+      for (int i = 0; i < NUM_SHARDS; i++) {
+        final Key shardKey = datastore.newKeyFactory().setKind(KIND_COUNTER_SHARD)
+            .newKey(counterId + "-" + i);
+        datastore.runInTransaction(transaction -> {
+          final Entity shard = transaction.get(shardKey);
+          if (shard == null) {
+            transaction.put(Entity.newBuilder(shard).set(PROPERTY_SHARD, 0).build());
+          }
+          return null;
+        });
+      }
+    }
+
+    /**
+     * Gets the counter state. Creates it in Datastore if it doesn't exist yet.
+     */
+    private static CounterSnapshot fromDatastore(Datastore datastore, String counterId) {
+      CounterSnapshot snapshot = new CounterSnapshot();
+
+      final Key limitKey = datastore.newKeyFactory().setKind(KIND_COUNTER_LIMIT).newKey(counterId);
+      final Entity limit = datastore.get(limitKey);
+      if (limit == null) {
+        snapshot.limit = Long.MAX_VALUE;
+        // Or IllegalStateException("No limit found in Datastore for " + counterId);?
+      } else {
+        snapshot.limit = limit.getLong(PROPERTY_LIMIT);
+      }
+
+      final EntityQuery queryShards = EntityQuery.newEntityQueryBuilder()
+          .setKind(KIND_COUNTER_SHARD)
+          .setFilter(CompositeFilter.and(
+              PropertyFilter.ge("__key__", counterId + "-0"),
+              PropertyFilter.lt("__key__", counterId + "-" + NUM_SHARDS)))
+          .setOrderBy(OrderBy.asc("__key__"))
+          .setLimit(NUM_SHARDS)
+          .build();
+      final QueryResults<Entity> shards = datastore.run(queryShards);
+      snapshot.shards = new ArrayList<Long>(NUM_SHARDS);
+      int shardCount;
+      for (shardCount = 0; shards.hasNext(); shardCount++) {
+        Entity shard = shards.next();
+        String shardKey = shard.getKey().toString();
+        int shardIndex = Integer.valueOf(shardKey.substring(shardKey.indexOf('-') + 1));
+        snapshot.shards.set(shardIndex, shard.getLong(PROPERTY_SHARD));
```

Comment:
As the `shards` list is empty (even though it has `NUM_SHARDS` capacity), calling `set()` will throw `IndexOutOfBoundsException`. 

Call `add()` instead.